### PR TITLE
Make the file writing code more flexible.

### PR DESCRIPTION
**Pull request by FreddieWitherden**
It is now possible for data sets to be specified at the time a file is being written as opposed to when the NativeWriter is constructed.

### CODE REVIEW

File: pyfr/plugins/base.py [hunk 1]
```diff
@@ -104,54 +104,29 @@ def _invoke_postaction(self, intg, **kwargs):
...
+            self._prepare_region_data_box(intg, *box)
         # All elements on a boundary
         else:
-            mdata = self._prepare_mdata_bcs(intg, fpdtype, nout, prefix,
-                                            region)
-            self._add_region_data = self._add_region_data_subset
-
-        # Construct the file writer
-        return NativeWriter(intg, mdata, basedir, basename)
+            self._prepare_region_data_bcs(intg, region)
```
Comment (GonzaloSaez) [inline pyfr/plugins/base.py:122]: `from pyfr.writers.native import NativeWriter` should be removed I think

File: pyfr/plugins/writer.py [hunk 2]
```diff
@@ -22,6 +28,9 @@ def __init__(self, intg, cfgsect, suffix=None):
         # Output field names
         self.fields = intg.system.elementscls.convarmap[self.ndims]
 
+        # Output data type
+        self.fpdtype = intg.backend.fpdtype
```
Comment (GonzaloSaez) [inline pyfr/plugins/writer.py:32]: Wouldn't it be interesting to give the users to output solution files in other than the default precision? (opt-in feature of course)

File: pyfr/plugins/base.py (hunk 1 above)
Comment (FreddieWitherden) [reply pyfr/plugins/base.py:121]: Fixed.

File: pyfr/plugins/writer.py (hunk 2 above)
Comment (FreddieWitherden) [reply pyfr/plugins/writer.py:32]: We can certainly add that without too much hassle.

**Pull request merged**